### ISSUE TITLE
Unexpected compile error: 'z' does not exist in type Record<"z" & "z", any>

### ISSUE THREAD

**Issue by kpdonn** [Microsoft/TypeScript#19595]
**TypeScript Version:**  Version 2.7.0-dev.20171029 (also affects 2.5)

**Code**

```ts
declare function functionGenerator<A extends string, B>(a: A[], b: B): 
    (arg: Record<A & keyof B, any> & {otherProp: any}) => void

const generatedFunction = functionGenerator(["z"], {z: ""})

generatedFunction({otherProp: "", z: ""})
```

**Expected behavior:**
Should compile successfully

**Actual behavior:**
`error TS2345: Argument of type '{ otherProp: string; z: string; }' is not assignable to parameter of type 'Record<"z" & "z", any> & { otherProp: any; }'.`
`Object literal may only specify known properties, and 'z' does not exist in type 'Record<"z" & "z", any> & { otherProp: any; }'.`

The `{otherProp: any}` part is just there because without it it seems to decide that `Record<"z" & "z", any>` is `any` and allows anything to be passed to `generatedFunction`.

This is the minimal reproduction example I was able to come up with. How I actually ran into this in practice was a lot more complicated but essentially boiled down to wanting something like `functionGenerator` to create a function that accepted an object where fields from A were required unless they were also keys of B where they would then be optional.

**Comment by andy-ms** [comment]
This behavior *does* work with `declare function f(arg: Record<"z" & "z", any> & { otherProp: any }): void;`, so we must not be correctly reporting the true type of `generatedFunction`.

**Comment by mhegazy** [comment]
we do not normalize intersections really. we should disable the error. but i think the right way to define this function is:
```ts
declare function functionGenerator<A extends string, B>(a: A[], b: B):
    (arg: Record<keyof B, any> & Record<A, any> & { otherProp: any }) => void;
```

**Comment by kpdonn** [comment]
Closing because I just checked and this stopped happening sometime between 2.6 and 2.7 so it must have been fixed as a side effect of other changes.
